### PR TITLE
feat: #1538 backend store user privilege change audit records

### PR DESCRIPTION
**Pull request by ianliuwk1019**
ref: #1538 
- create a service as main logic place for storing audit records (grant/revoke).
- create a repository to save audit records.
- when storing with scoped with forest_client, also record the forest client name.
- new api-client gen code.
- some minor refactoring.

Will do tests in separate pr.

Sorry, lots of generated api-client code, can ignore them. Also quite some files with only imports reorganized by IDE, can ignore them as well.
<html xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/po-yi.liu/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/po-yi.liu/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">

</head>

<body link="#467886" vlink="#96607D">

privilege_change_audit_id | application_id | change_date | change_performer_user_details | change_performer_user_id | change_target_user_id | create_date | create_user | privilege_change_type_code | privilege_details
-- | -- | -- | -- | -- | -- | -- | -- | -- | --
16 | 2 | 2024-09-19 21:06:10.816231+00 | {"email": "a@b.com", "username":   "IANLIU", "last_name": "Liu",   "first_name": "Ian"} | 4 | 9 | 2024-09-19 21:06:10.816228+00 | IANLIU | GRANT | {"roles":   [{"role": "Reviewer"}], "permission_type":   "End User"}
17 | 2 | 2024-09-19 21:07:40.022031+00 | {"email": "a@b.com", "username":   "IANLIU", "last_name": "Liu",   "first_name": "Ian"} | 4 | 9 | 2024-09-19 21:07:40.022029+00 | IANLIU | GRANT | {"roles":   [{"role": "Submitter", "scopes":   [{"client_id": "00001011", "scope_type":   "Client", "client_name": "AKIECA EXPLORERS   LTD."}, {"client_id": "00001012",   "scope_type": "Client", "client_name":   "BELL POLE COMPANY"}]}], "permission_type": "End   User"}
18 | 2 | 2024-09-19 21:12:00.807222+00 | {"email": "a@b.com", "username":   "IANLIU", "last_name": "Liu",   "first_name": "Ian"} | 4 | 9 | 2024-09-19 21:12:00.807219+00 | IANLIU | REVOKE | {"roles":   [{"role": "Submitter", "scopes":   [{"client_id": "00001012", "scope_type":   "Client", "client_name": "BELL POLE   COMPANY"}]}], "permission_type": "End User"}
19 | 2 | 2024-09-19 21:12:27.511049+00 | {"email": "a@b.com", "username":   "IANLIU", "last_name": "Liu",   "first_name": "Ian"} | 4 | 9 | 2024-09-19 21:12:27.511047+00 | IANLIU | REVOKE | {"roles":   [{"role": "Reviewer"}], "permission_type":   "End User"}

</body>

</html>

### CODE REVIEW

File: server/admin_management/api/app/integration/forest_client_integration.py [hunk 1]
```diff
@@ -18,6 +18,12 @@ class ForestClientIntegrationService():
     Spec of API:
         test: https://nr-forest-client-api-test.api.gov.bc.ca/
         prod: https://nr-forest-client-api-prod.api.gov.bc.ca/
+
+    Note! This is external API integration and FAM supports 3 applications environments in PROD.
+          For FAM(PROD)-Application(PROD): it will connect to ForestClientAPI PROD instance.
+          For rest of application environments (TEST/DEV) in FAM(PROD): it will use TEST instance.
+          For FAM environment management relating to the use of external API,
+          see ref @FAM Wiki: https://github.com/bcgov/nr-forests-access-management/wiki/Environment-Management
     """
```
Comment (ianliuwk1019) [inline server/admin_management/api/app/integration/forest_client_integration.py:27]: Nothing change at server/admin_management.
Only copy same comment from server/backend for forest_client_integration.py.

File: server/backend/api/app/crud/services/permission_audit_service.py [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+from fastapi import HTTPException
+from sqlalchemy.orm import Session
+
+LOGGER = logging.getLogger(__name__)
+
+
+class PermissionAuditService:
+
+    def __init__(self, db: Session):
+        self.repo = PermissionAuditRepository(db)
```
Comment (ianliuwk1019) [inline server/backend/api/app/crud/services/permission_audit_service.py:35]: This is `server/backend`. Everything was under `crud` folder.
In the past, we always liked to get rid of the `crud` but had no time and would be too big of refactoring at once.
For this new feature, I started using same layers structure (server/repository) from `server/admin_management` to start moving toward consistency with `server/admin_management`.

File: server/backend/api/app/crud/services/permission_audit_service.py [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+
+    def store_user_permissions_granted_audit_history(
+        self,
+        requester: RequesterSchema,
+        change_target_user: FamUser,
+        new_user_permission_grated_list: List[FamUserRoleAssignmentCreateRes]
+    ):
+        success_granted_list = list(filter(
+            lambda res: res.status_code == HTTPStatus.OK, new_user_permission_grated_list
+        ))
```
Comment (ianliuwk1019) [inline server/backend/api/app/crud/services/permission_audit_service.py:45]: Current FAM implementation can grant a user with 1 role with multiple forest_client and each has its creation response with possible different return code. So it is a list to be filtered for the success ones.

File: server/backend/api/app/crud/services/permission_audit_service.py [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+    @staticmethod
+    def to_change_performer_user_details(
+        requester: RequesterSchema
+    ) -> PrivilegeChangePerformerSchema:
+        return PrivilegeChangePerformerSchema(
+            username=requester.user_name,
+            first_name=requester.first_name,
+            last_name=requester.last_name,
+            email=requester.email
+        )
```
Comment (ianliuwk1019) [inline server/backend/api/app/crud/services/permission_audit_service.py:95]: This is only convenient method for transformation between schema objects.

File: server/backend/api/app/crud/services/permission_audit_service.py [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+        scopes = list(map(__map_to_privilege_role_scope, enduser_privliege_list))
+        return PrivilegeDetailsSchema(
+            permission_type=PrivilegeDetailsPermissionTypeEnum.END_USER,
+            roles=[
+                PrivilegeDetailsRoleSchema(
+                    role=enduser_privliege_list[0].detail.role.display_name,
+                    scopes=scopes
+                )
+            ]
+        )
```
Comment (ianliuwk1019) [inline server/backend/api/app/crud/services/permission_audit_service.py:122]: This is only convenient method for transformation between schema objects to build the end user `enduser_privliege_granted_details`

File: server/backend/api/app/crud/services/permission_audit_service.py [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+
+        return PrivilegeDetailsSchema(
+            permission_type=PrivilegeDetailsPermissionTypeEnum.END_USER,
+            roles=[
+                PrivilegeDetailsRoleSchema(
+                    role=revoked_permission_role.display_name,
+                    scopes=[revoke_privilege_details_scope]
+                )
+            ]
+        )
```
Comment (ianliuwk1019) [inline server/backend/api/app/crud/services/permission_audit_service.py:172]: This is only convenient method for transformation between schema objects to build the end user `enduser_privliege_revoked_details`

File: server/backend/api/app/integration/forest_client_integration.py [hunk 7]
```diff
@@ -8,7 +8,7 @@
 LOGGER = logging.getLogger(__name__)
 
 
-class ForestClientService():
+class ForestClientIntegrationService():
```
Comment (ianliuwk1019) [inline server/backend/api/app/integration/forest_client_integration.py:11]: Renamed this service name to be the same name from the same service under: `server/admin_management`. So to keep them consistent. (we had two copies of the same service).

File: server/backend/api/app/schemas/__init__.py [hunk 8]
```diff
@@ -56,8 +52,3 @@
 """
 from .requester import RequesterSchema
 from .target_user import TargetUserSchema
-
-# ---------- Permission Audit History Schemas ---------- #
-from .privilege_details import PrivilegeDetailsSchema
-from .privilege_change_performer import PrivilegeChangePerformerSchema
-from .permission_audit_history import PermissionAduitHistoryRes
```
Comment (ianliuwk1019) [inline server/backend/api/app/schemas/__init__.py:63]: Didn't change this file, but for some reason in I touched it my IDE rearrange the imports on save.
Maybe we can remove the comment "#---#" from here, no need to group it just let IDE do the job...

File: server/backend/api/app/schemas/fam_role_with_client.py [hunk 9]
```diff
@@ -1,31 +1,21 @@
...
-        fields={
-            "update_user": {"exclude": True},
-            "role_purpose": {"exclude": True},
-            "parent_role_id": {"exclude": True},
-            "application_id": {"exclude": True},
-            "forest_client_number": {"exclude": True},
-            "role_id": {"exclude": True},
-            "create_user": {"exclude": True},
-        },
-    )
```
Comment (ianliuwk1019) [inline server/backend/api/app/schemas/fam_role_with_client.py:31]: No need for these, removed.

File: server/backend/api/app/schemas/__init__.py [hunk 10]
```diff
@@ -1,46 +1,42 @@
 # --------------------------------- FAM Application --------------------------------- #
```
Comment (ianliuwk1019) [inline server/backend/api/app/schemas/__init__.py:1]: Didn't change this file, but for some reason when I touched it my IDE rearranged the imports on save. Please ignore them.
Maybe we can remove the comment "#---#" from here, no need to group it just let IDE do the job...

**Comment by basilv** [comment]
Looking at the sample content in the table, I see JSON in the role column: {"roles": [{"role": "Reviewer", "scopes": null}], "permission_type": "End User"}. I'd expect the scopes attribute to be omitted entirely if there is no scope, not set to null.

**Comment by basilv** [comment]
I don't have time to detailed review all the code changes, but the general refactoring done makes  sense to me.

**Comment by ianliuwk1019** [comment]
> Looking at the sample content in the table, I see JSON in the role column: {"roles": [{"role": "Reviewer", "scopes": null}], "permission_type": "End User"}. I'd expect the scopes attribute to be omitted entirely if there is no scope, not set to null.

Your time is much appreciated, thanks for noticing that Basil!! I had the same wondering too but due to it is a "**key**" of the object (in python schema) initially I thought it probably cannot be omit entirely. But... not sure if there is a way..., let me look into it.

**Comment by ianliuwk1019** [comment]
> > Looking at the sample content in the table, I see JSON in the role column: {"roles": [{"role": "Reviewer", "scopes": null}], "permission_type": "End User"}. I'd expect the scopes attribute to be omitted entirely if there is no scope, not set to null.
> 
> Your time is much appreciated, thanks for noticing that Basil!! I had the same wondering too but due to it is a "**key**" of the object (in python schema) initially I thought it probably cannot be omit entirely. Also, if that's the case (to omit entirely), I am also wondering if it will cause problem when JSON is being deserialized back to object. But... not sure if there is a way..., let me look into it.

Hi @basilv , after some digging, there is no easy way for conditionally removing the attribute out of Pydantic schema object using Pydantic currently. However, by using python "del" to delete the attribute (when None) does the trick. And the JSON example is updated. Now it is showing `{"roles": [{"role": "Reviewer"}], "permission_type": "End User"}`

File: server/backend/api/app/crud/crud_user_role.py [hunk 11]
```diff
@@ -134,35 +137,44 @@ def create_user_role_assignment_many(
...
+    )
 
-    LOGGER.debug(f"User/Role assignment executed successfully: {create_return_list}")
-    return create_return_list
+    return new_user_permission_granted_list
 
 
 def create_user_role_assignment(
-    db: Session, user: models.FamUser, role: models.FamRole, requester: str
+    db: Session, user: models.FamUser, role: models.FamRole, requester_cognito_user_id: str  #cognito_user_id,
```
Comment (craigyu) [inline server/backend/api/app/crud/crud_user_role.py:168]: Remove the commented out code `#cognito_user_id`

**Review by craigyu — APPROVED** [review]
Really good work overall 💯, just left a lint change request

File: server/backend/api/app/crud/crud_user_role.py (hunk 11 above)
Comment (ianliuwk1019) [reply server/backend/api/app/crud/crud_user_role.py]: Ah~ thanks for catching that! Removed.

**Pull request merged**